### DISCUSSION TITLE
The best practice for entity removal inside lifecycle hooks?

### DISCUSSION THREAD

**Discussion by pawelsmolenski** [mikro-orm/mikro-orm#4028] [Q&A] [answered]
What is the best practice for entity removal inside lifecycle hooks?

I am performing some database updates on lifecycle events (e.g. onFlush). I have already learned how to create and mark for insertion new entities inside those hooks, but how to delete arbitrary ones which I am considering obsolete?

Elaborating with a pseudo-code: 

Having, let's say, two unrelated or loosely related entities A and B:
```js
@Entity()
class A {
    @PrimaryKey()
    id: string
    
    @Field()
    sth: boolean;
}

@Entity()
class B {
    @PrimaryKey()
    id: string
    
    @Field()
    whatever: string;
}
```

and a subscriber service:

```js
class ASubscriber implements EventSubscriber<A> {
    public async onFlush(args) {
        args.uow.getChangeSets()
            .filter((changeSet: ChangeSet<AnyEntity>) => changeSet.entity instanceof A)
            .forEach((changeSet) => {
                
                // todo: delete some instance of B here
            
            });
    }
}
```

How to delete arbitrary B entity on a update of an entity A within a single transaction:

```js
em.begin();

const a = await em.getRepository(A).findOrFail({ id: "4de9df02-9e85-4c46-89fc-0efd040e3665" });
a.sth = true;
em.flush();

// test of the lifecycle hook
em.getRepository(B).findAll() // arbitrary instance of B entity should be already removed here

em.commit();
```

?

**Comment by B4nan** [comment, marked as answer]
Hmm I guess this one deserves some shortcut, you could create the changeset manually I guess. Note you can't do this with `args.uow.getChangeSets()` as that gives you a copy, the real changesets are stored in a map in a private property of UoW.

```ts
(uow as any).changeSets.set(entity, new ChangeSet(entity, ChangeSetType.DELETE, {}, helper(entity).__meta));
```

**Comment by pawelsmolenski** [comment]
Thank you B4nan. Given solution works perfectly for me. It seems, that MikroORM still holds many secrets for me to unveil.
